Port Ryu float-to-string conversion algorithm to Scala

tanishiking opened the pull request:
close https://github.com/scala-wasm/scala-wasm/issues/11

Implement the Ryu algorithm for converting binary floating-point numbers the decimal string representation, as described in "Ryu: fast float-to-string conversion" by Ulf Adams (PLDI 2018). https://dl.acm.org/doi/10.1145/3192366.3192369

This port is based on the reference Java implementation from https://github.com/ulfjack/ryu with some comments and changes.

Most of the changes are addition of `RyuDouble` and `RyuFloat`, and Emitter to use those function for `toStringForConcat`. Rest of the change is adding a big pre-computed table, and test cases.

tanishiking committed:
Port RyuDouble

Implement the Ryu for converting binary floating-point numbers the decimal string representation, as described in "Ryu: fast float-to-string conversion" by Ulf Adams (PLDI 2018).
https://dl.acm.org/doi/10.1145/3192366.3192369

This port is based on the reference Java implementation from https://github.com/ulfjack/ryu with some comments and changes.

We use `RyuDouble` both for Float and Double because of Scala.js spec.

We do not constant fold on JVM on purpose because
the rounding behavior of Double.toString differs between versions before and after JDK 19.
For instance, in JDK 21, `Double.toString(double)` specifies "round to even" for boundary cases,
whereas this specific boundary behavior seems not specified in JDK 17.
https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Double.html#toString(double)
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html#toString(double)

tanishiking commented:
@sjrd Could you take a quick look especially on changes under `linker` and `test-suite`?
The core implementation is ported from https://github.com/ulfjack/ryu with some refactoring + comments on it.

plokhotnyuk commented:
@tanishiking Have you considered to use more efficient algorithm from Raffaello Giulietti?

"The Schubfach way to render doubles":
https://drive.google.com/file/d/1luHhyQF9zKlM8yJ1nebU0OgVYhfC6CBN/view

Java sources with the license: 
https://github.com/c4f7fcce9cb06515/Schubfach/tree/3c92d3c9b1fead540616c918cdfef432bca53dfa/todec/src/math

Scala implementation that is based on it:
https://github.com/plokhotnyuk/jsoniter-scala/blob/c60560deb71a5f24b99d395f4b823569aa05a601/jsoniter-scala-core/jvm-native/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L2238-L2536

BTW, this algorithm is already implemented in the latest versions of JDK

tanishiking commented:
Hi @plokhotnyuk,
I reviewed papers for Dragon4, Grisu3, and Ryu, and chose Ryu for its completeness, and since it's used in `scala-native` as well. I didn't seriously consider more recent algorithms like Schubfach and Dragonbox. 😞 
However, after looking at `jsoniter-scala`, the Java implementation, and take a quick look at the paper, Schubfach seems to be a very elegant algorithm and would definitely be worth adopting (whereas Dragonbox seems somewhat more complicated).
And indeed, Schubfach seems to be faster than Ryu https://github.com/abolz/Drachennest

That said, performance isn't our top priority at the moment, so I feel like continuing with tried and tested Ryu for now and consider migrating to Schubfach in the future (perhaps in scala-native as well 🙌 ).

Anyway, thanks for sharing @plokhotnyuk !

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -2322,16 +2322,18 @@ private class FunctionEmitter private (
-              fb += wa.Drop
-              fb ++= ctx.stringPool.getConstantStringInstr("0")
+              fb += wa.Call(genFunctionID.forMethod(MemberNamespace.PublicStatic,
+                  SpecialNames.RyuFloatClass, SpecialNames.floatToStringMethodName))
```
By spec, we must use `doubleToString` even for floats. Otherwise `x.toString()` won't give the same result as `(x: Any).toString()`.

So we really don't need the variant for `Float`s. We only use the one for `Double`s.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala:
```diff
@@ -587,7 +587,11 @@ object Emitter {
+
+      // TODO Ideally we should not require these, but rather adapt to their absence
+      callStaticMethod(RyuFloatClass, floatToStringMethodName),
+      callStaticMethod(RyuDoubleClass, doubleToStringMethodName)
```
These should be in the `linker-private-library`, like `fmod`. And those requirements (in fact just the `double` one, see my other comment) should be conditioned on `targetPureWebAssembly`. You can use `cond` for that, like we do for error classes 20 lines above.

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -63,32 +63,271 @@ class DoubleTest {
+     * https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Double.html#toString(double)
+     * https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html#toString(double)
+     * To ensure stability, we are skipping comparisons against `value.toString()`, to avoid being depend on the
+     * specific JDK's string conversion implementation.
```
This comment, in fact, explains why it's invalid to constant-fold `floatLit.toString()` on the JDK. We can't do that.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -4212,17 +4212,30 @@ private[optimizer] abstract class OptimizerCore(
-        jsNumberToString(value).fold(preTrans)(s => constant(s))
+      def forFloat(value: Float): PreTransform = {
+        if (targetPureWasm) {
+          constant(value.toString)
```
As the ignored test demonstrates, this is an invalid optimization. We must never constant-fold using the JDK's `floatValue.toString()`.

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -63,32 +63,271 @@ class DoubleTest {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
+  @Test def toStringTest(): Unit = {
+    assumeTrue(
+        "The output of JS and JVM (depends on JDK version) differs slightly from the Ryu implementations.",
```
Could we tweak Ryu's boundary conditions to always do the same as JS? JS has been well-specified since forever, and that's already what Scala.js-on-Wasm uses.

sjrd commented on javalib/src/main/scala/java/lang/flttodec/RyuDoubleTables.scala:
```diff
@@ -0,0 +1,720 @@
+
+  /** Pow5Split contains pre-computed 5^q values split into 4 31-bit parts. */
+  val Pow5Split = Array(
+    ( 134217728,          0,          0,          0),
```
This boxes all the ints. We don't want that here.

We could define a `class FourInts` with 4 `Int` fields (anyway you won't have `Tuple`s available in the linker-private-library). But even that will create a lot of unnecessary code in the generated Wasm.

Consider using a single, flat `Array[Int]`, and index into with `index * 4 + part`.

sjrd commented on javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala:
```diff
@@ -0,0 +1,446 @@
+
+  @inline
+  private def printString(output0: Long, olength: Int, exp0: Int,
+      sign: Boolean, scientificNotation: Boolean): String = {
```
```suggestion
  private def printString(output0: Long, olength: Int, exp0: Int,
      sign: Boolean, scientificNotation: Boolean): String = {
```
This large method does not return a tuple, so I guess it's counter-productive to inline it.

tanishiking replied:
Since `printString` is called only from `doubleToString`, I guess it's not counter-productive I guess?
(The original Ryu code has inlined into the `doubleToString`, but I split into a function for readability) (embedding the printString into `doubleToString` is the best for performance maybe though)

sjrd commented on javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala:
```diff
@@ -0,0 +1,446 @@
+      v /= 5
+      count += 1
+    }
+    throw new IllegalArgumentException(s"$value")
```
`AssertionError`? Same for the other ones. If we run into an exception anywhere in this conversion, it's a bug in the conversion, not a user-space mistake.

tanishiking commented:
Thanks @sjrd for review!
Could you take another look if all feedbacks are fixed?

- Use `doubleToString` even for `Float`
- Don't constant fold in JVM
- Move RyuDouble to `linker-private-library` and requires it only when targeting pure Wasm.
- Use flat `Array[Int]`
- Align the output of Ryu with JS.
  - mostly about scientific notation behavior, and trim trailing `.0`.
- **Made ToStringTest to test roundtrip as well**.

Maybe you don't need to take a look RyuDouble implementation very carefully, since it's basically a port from reference implementation, and maybe we swap the implementation to another algorithm in future (?)

sjrd commented on javalib/src/main/scala/java/util/internal/Tuples.scala:
```diff
@@ -21,6 +21,10 @@ final class Tuple3[+T1, +T2, +T3](val _1: T1, val _2: T2, val _3: T3)
 
+@inline
+final class Tuple6[+T1, +T2, +T3, +T4, +T5, +T6](
+    val _1: T1, val _2: T2, val _3: T3, val _4: T4, val _5: T5, val _6: T6)
```
This is now unused.

sjrd commented on linker/jvm/src/main/scala/org/scalajs/linker/backend/emitter/PrivateLibHolder.scala:
```diff
@@ -31,6 +31,8 @@ object PrivateLibHolder {
       "org/scalajs/linker/runtime/WasmRuntime.sjsir",
       "org/scalajs/linker/runtime/WasmRuntime$.sjsir",
+      "org/scalajs/linker/runtime/RyuDouble.sjsir",
+      "org/scalajs/linker/runtime/RyuDouble$.sjsir",
```
Put those between `RuntimeLong` and `UndefinedBehaviorError` to preserve the alphabetical order.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -4221,8 +4221,8 @@ private[optimizer] abstract class OptimizerCore(
-        case FloatLiteral(value)   => forFloatingPoint(value)
-        case DoubleLiteral(value)  => forFloatingPoint(value)
+        case FloatLiteral(value)  if !targetPureWasm => forFloatingPoint(value)
+        case DoubleLiteral(value) if !targetPureWasm => forFloatingPoint(value)
```
This change is probably (hopefully) not necessary anymore, is it?

tanishiking replied:
Yeah, now the RyuDouble basically prints the same as JavaScript, we can remove this 👍

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
     if (executingInJVM) {
       assertEquals("0.0", 0.0.toString)
-      assertEquals("-0.0", (-0.0).toString)
+      assertEquals("-0.0", -0.0.toString)
```
Revert these changes to preserve the parens?

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
+      assertEquals(expectedStr, actualStr)
+      // Test roundtrip: parsing the string should give back the exact same value
+      val parsed = JDouble.parseDouble(actualStr)
+      assertEquals(value, parsed, 0.0)
```
Use `org.scalajs.testsuite.utils.AssertExtensions.assertExactEquals` here.

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
+
+    // boundaryConditions
+    // x = 1.0E7
+    test("10000000", 1.0E7d)
```
Nit: be consistent with the case of the `e`s. In this codebase we prefer the lowercase version (notably because that's what `toString()` does).

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
+    test("1.7976931348623157e+308", JDouble.longBitsToDouble(0x7fefffffffffffffL))
+    // TODO: (align the output between Ryu and JS)
+    assertEquals(
+      if (executingInPureWebAssembly) "4.9e-324" else "5e-324",
```
Can we do that now, before we forget?

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala:
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
+    // 32-bit opt-size=1:  30 <= dist <= 50
+    // 64-bit opt-size=0:  50 <= dist <= 50
+    // 64-bit opt-size=1:  30 <= dist <= 50
+    assertEquals("1.7800590868057611e-307", ieeeParts2Double(sign = false, 4, 0).toString)
```
```suggestion
    test("1.7800590868057611e-307", ieeeParts2Double(sign = false, 4, 0))
```
and everywhere below?

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/FloatTest.scala:
```diff
@@ -73,25 +73,23 @@ class FloatTest {
-  @Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
-        executingInPureWebAssembly)
+@Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
```
```suggestion
  @Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
```

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/FloatTest.scala:
```diff
@@ -73,25 +73,23 @@ class FloatTest {
     if (executingInJVM) {
       assertEquals("0.0", 0.0f.toString)
-      assertEquals("-0.0", (-0.0f).toString)
+      assertEquals("-0.0", -0.0f.toString)
```
Same here as above: revert those changes?

sjrd commented on linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala:
```diff
@@ -0,0 +1,1162 @@
+
+  @inline
+  private def convertBinaryToDecimalInterval(u: Long, v: Long, w: Long, e2: Int,
+      isGeneralLowerBoundCase: Boolean, even: Boolean): (Long, Long, Long, Int, Boolean, Boolean) = {
```
Ouch, this shouldn't work. I think the `JavalibIRCleaner` is too lenient in the linker-private-library. It shouldn't allow references to our custom tuple classes from the javalib. The linker-private-library should always be able to link with any ABI-compliant version of the javalib.

Use a custom `@inline private final class` instead of the 6-tuple.

sjrd commented on linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala:
```diff
@@ -0,0 +1,1162 @@
+
+  @inline
+  private def computeShortest(aq0: Long, bq0: Long, cq0: Long, za0: Boolean, zb0: Boolean,
+      e10: Int, even: Boolean, scientificNotation: Boolean): (Long, Int) = {
```
Same here. Use a custom `@inline private final class` for the tuple.

sjrd reviewed (changes requested):
Mostly nits.

There's a problem with the use of tuples in the linker-private-library.

There's also the case of `4.9` vs `5`. If it's not too much trouble, we should fix it now before we forget. If it's hard, let's open an issue to keep track of it.

tanishiking committed:
Align Ryu output with JS semantics for scientific notation

JDK's Double.toString prints at least two digits in scientific
notation output. For example, `Double.longBitsToDouble(1).toString()`
prints `4.9E-324`, while JavaScript prints `5e-324`.

JavaScript has no such requirement for scientific notation and
always prints the shortest digits.
https://tc39.es/ecma262/#sec-numeric-types-number-tostring

This commit aligns the Ryu output with JS semantics.

tanishiking commented:
> There's also the case of 4.9 vs 5. If it's not too much trouble, we should fix it now before we forget. If it's hard, let's open an issue to keep track of it.

Aligned the output with the JS in https://github.com/scala-wasm/scala-wasm/pull/116/commits/0ca3688debbacdaca9198fc4c89d034ef535ca99
It was just `RyuDouble#computeShortest` had a logic to keep at least two digits in scientific notation output because of JDK `Double.toString` does. Just removed that (because JS always print the shortest digits) :+1:

sjrd reviewed (approved):
Nice!

plokhotnyuk commented:
@tanishiking [Today I adopted the XJB algorithm](https://github.com/plokhotnyuk/jsoniter-scala/commit/f7f48455ebf2b026e4511b3a6c5fcb816d538ceb). In practice this noticeably reduced CPU usage and sped up float/double serialization comparing to previous Schubfach-base implementation (up to 25% on JVM and up to 45% in JS browsers on real-world message samples).

tanishiking commented:
@plokhotnyuk 
Wow, it's an algorithm that was just announced in a preprint recently? https://github.com/xjb714/f2dec_bench

I haven't checked the algorithm yet, but if it handles float-to-string conversion correctly, that's incredible 😮 (It's even faster than DragonBox) (and 25,45% speed up in JVM and JS is awesome!).

I'm curious that, from what I read in https://internals.rust-lang.org/t/algorithm-new-float-double-to-string-algorithm/23670/, it doesn't seem like they have verified the algorithms properties Information Preservation (round-tripping from decimal back to binary and vice versa), Minimum length, and Correct Rounding. Did you find any problem, or is it working good so far?

plokhotnyuk commented:
> Did you find any problem, or is it working good so far?
I have used brute force and checked all `floats` - the output is always parsed without any bit of loss and representation is always p 



> @plokhotnyuk Wow, it's an algorithm that was just announced in a preprint recently? https://github.com/xjb714/f2dec_bench
>

Yes, I only skimmed through the preprint and switched to adopt and test it first ;)
 
> I haven't checked the algorithm yet, but if it handles float-to-string conversion correctly, that's incredible 😮 (It's even faster than DragonBox) (and 25,45% speed up in JVM and JS is awesome!).
> 
> I'm curious that, from what I read in https://internals.rust-lang.org/t/algorithm-new-float-double-to-string-algorithm/23670/, it doesn't seem like they have verified the algorithms properties Information Preservation (round-tripping from decimal back to binary and vice versa), Minimum length, and Correct Rounding. Did you find any problem, or is it working good so far?

I exhaustively tested all 2^32 possible float values: every one round-trips perfectly with no bit loss, and the textual representation is always the same length or shorter than Java’s Schubfach implementation in the latest JDKs. I also spent many CPU-hours stress-testing double values, and so far everything looks rock solid.

tanishiking commented:
Sounds very promising!
I'll keep an eye on the project, and we might want to move from ryu to XJB in the future :) Thank you so much for sharing 🙏

tanishiking commented:
Sounds very promising!
I'll keep an eye on the project, and we might want to move from ryu to XJB in the future :) Thank you so much for sharing 🙏
